src/io/flutter/ProjectOpenActivity.java

```diff
@@ -35,12 +54,73 @@
 public class ProjectOpenActivity implements StartupActivity, DumbAware {
   public static final ProjectType FLUTTER_PROJECT_TYPE = new ProjectType("io.flutter");
   private static final Logger LOG = Logger.getInstance(ProjectOpenActivity.class);
+  private static final String DOWNLOAD_PATH = FileUtils.platformPath();
 
   public ProjectOpenActivity() {
   }
 
+  private void verifyJxBrowser(Project project) {
+    final File directory = new File(DOWNLOAD_PATH);
+    if (!directory.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      directory.mkdirs();
+    }
+
+    String jxbrowserFileName = JxBrowserUtils.getPlatformFileName();
+    final File jxbrowserPlatformFile = new File(DOWNLOAD_PATH + File.separator + jxbrowserFileName);
+    if (jxbrowserPlatformFile.exists()) {
+      // Skip downloading
+      System.out.println("JxBrowser platform file already exists");
+      loadClasses(jxbrowserPlatformFile);
+      return;
+    }
+
+    DownloadableFileService service = DownloadableFileService.getInstance();
+    DownloadableFileDescription description = service.createFileDescription(JxBrowserUtils.getDistributionLink(jxbrowserFileName), jxbrowserFileName);
+    FileDownloader downloader = service.createDownloader(Collections.singletonList(description), jxbrowserFileName);
+
+    Task.Backgroundable task = new Task.Backgroundable(project, "Downloading jxbrowser") {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        try {
+          List<Pair<File, DownloadableFileDescription>> pairs = downloader.download(new File(DOWNLOAD_PATH));
+          Pair<File, DownloadableFileDescription> first = ContainerUtil.getFirstItem(pairs);
+          File file = first != null ? first.first : null;
+          if (file != null) {
+            System.out.println("File downloaded: " + file.getAbsolutePath());
+          }
+          loadClasses(file);
+        }
+        catch (IOException e) {
+          System.out.println("Unable to download file");
+        }
+      }
+    };
+    BackgroundableProcessIndicator processIndicator = new BackgroundableProcessIndicator(task);
+    processIndicator.setIndeterminate(false);
+    ProgressManager.getInstance().runProcessWithProgressAsynchronously(task, processIndicator);
+  }
+
+  private void loadClasses(File file) {
```

Comment:
I think we should do a little refactoring. See AddToAppUtils.getStaticFieldValue(). If you move that method to, say, SystemUtils then you can just use it here. Unfortunately, I defined it in a class that's part of flutter-studio so you can't use it directly.

Reply:
Should this function return the `addUrl` method? Something like `Method method = SystemUtils.getStaticFieldValue(URLClassLoader.class, Function<String, void>.class, "addUrl");` to replace everything before `method.invoke(classLoader, url);` ?

Reply:
Hmm, I don't know what I was thinking. See various methods in ReflectionUtils for helpers that do some of what you're doing here. Please ignore AddToAppUtils.